File: OS_A01/sqrt.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////
// SquareRoot C++ Program
// This program spawns children, who will help to determine square roots.
// To compile:  g++ sqrt.cpp -o sqrt
// To run:  ./sqrt > sqrt.out
// To monitor system resources: (Linux): Applications->System Tools->System Monitor->Resources
// Use Spreadsheet Applications->Office->LibreOffice Calc to show efficiency of multiple threads.
//////////////////////////////////////////////////////////////////////////////////////////////////////////////
#include <iostream>
#include <stddef.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <ctime>
#include <stdlib.h>
//#include <processthreadsapi.h>   // Windows: GetCurrentProcessorNumber()
#include <utmpx.h>
#include <math.h>
using namespace std;

#define numChild  6
#define total  100     // 200000000
double global=0.0;

class SquareRoot
{
public:
    SquareRoot() { memAttr=0.0; }
    void child(double begin, double end);
private:
    int start_s;     // start time
    double memAttr;
};

int main()
{
    int i;
    double range = total/numChild;
    double begin=5999901.0;
    int stat;

    cerr << "Run SquareRoot " << total << ":" << numChild << endl;
    SquareRoot Sqrt;

    // Spawn children processes
    for (i=0; i<numChild; i++) {
        if (fork() == 0) Sqrt.child(begin, begin+range);
        begin += range + 1;
    }

    // Wait for children to finish
    for (i=0; i<numChild; i++)
        wait(&stat);
    cerr << "All Children Done: " << numChild << endl;
}

void SquareRoot::child(double begin, double end) {
    // Print the current CPU number
    //cerr << "CPU:"<< GetCurrentProcessorNumber() << endl; // Windows 
    //cerr << "CPU=" << sched_getcpu() << endl;               // Linux version

    // Get the current time
    start_s = clock();
    double totalSum = 0.0;
    for (int local = begin; local < end; local++) {
        double root = sqrt(local);
        // revised lines to do prints:
        cout << local << ":" << root << " ";
        if (local%5==0) cout << endl;
        totalSum += root;
    }

    // Are class attributes and globals shared?
    cerr << "  totalSum=" << totalSum << " global=" << ++global << " memAttr=" << ++memAttr << endl;
    // Calculate execution time in ms and print
    int stop_s = clock();
    cerr << "time: " << (stop_s - start_s) / double(CLOCKS_PER_SEC) * 1000 << endl;
    exit(0);
}

File: Relevant_School_Coding_Projects/OS_A01_Files/Sqrt.java
import java.util.concurrent.TimeUnit;

public class Sqrt implements Runnable {
    private double begin;
    private double end;
    private long start_s;

    public Sqrt(double begin, double end) {
        this.begin = begin;
        this.end = end;
    }

    public static void main(String[] args) throws InterruptedException {
        
        int numChild = 4;
        int total = 100;
        double range = total / numChild;
        double begin = 5999901.0;

        System.err.println("Run Sqrt " + total + ":" + numChild);

        // Spawn child threads
        Thread[] threads = new Thread[numChild];
        for (int i = 0; i < numChild; i++) {
            threads[i] = new Thread(new Sqrt(begin, begin + range));
            threads[i].start();
            begin += range + 1;
        }

        // Wait for child threads to finish
        for (int i = 0; i < numChild; i++) {
            threads[i].join();
        }

        System.err.println("All Children Done: " + numChild);
    }

    public void run() {
        // Print the current thread ID
        System.err.println("Thread ID=" + Thread.currentThread());

        // Calculate and print square roots
        for (double i = begin; i <= end; i++) {
            double sqrt = Math.sqrt(i);
            System.out.println(i + ": " + sqrt);
        }

        // Calculate execution time in ms and print
        System.err.println("time: " + TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - start_s));

    }

    
}